vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,456 @@
+            let base_offset = size_of::<AdvancedLoggerMessageEntryV2>()
+                .checked_add(message_len as usize)
+                .ok_or_else(|| {
+                    ProcessingError(format!(
+                        "Overflow: AdvancedLoggerMessageEntryV2 size ({}) + message_len ({})",
+                        size_of::<AdvancedLoggerMessageEntryV2>(),
+                        message_len
+                    ))
+                })?;
+
+            // Add padding for 8-byte alignment
+            let aligned_offset = base_offset.checked_add(ALIGNMENT_MASK).ok_or_else(|| {
+                ProcessingError(format!(
+                    "Overflow: base_offset ({}) + {}",
+                    base_offset, ALIGNMENT_MASK
+                ))
+            })?;
+            let next_offset = aligned_offset & !ALIGNMENT_MASK;
+
+            // Update overall bytes read counter
+            bytes_read = bytes_read.checked_add(next_offset).ok_or_else(|| {
+                ProcessingError(format!(
+                    "Overflow: bytes_read ({}) + next_offset ({})",
+                    bytes_read, next_offset
+                ))
+            })?;
+
+            // Advanced to the next entry with boundary checks
+            if next_offset >= buffer_slice.len() {
+                // We have reached the end of the buffer
+                break;
+            }
+            buffer_slice = &buffer_slice[next_offset..];
+        }
+
+        // Process remaining messages
+        if is_accumulating {
+            logs.push(EfiDiagnosticsLog {
+                debug_level,
+                time_stamp,
+                phase,
+                message: std::mem::take(&mut accumulated_message)
+                    .trim_end_matches(&['\r', '\n'][..])
+                    .to_string(),
+            });
+            entries_processed += 1;
+        }
+
+        // Print summary statistics
+        tracelimit::info_ratelimited!(
+            "EFI Diagnostics: Processed {} entries, Read {} bytes",
+            entries_processed,
+            bytes_read
+        );
+
+        Ok(())
+    }
+}
+
+impl UefiDevice {
+    pub(crate) fn process_diagnostics(&mut self, gpa: u32, gm: GuestMemory) {
+        // Do not process if already done
+        if self.processed_diagnostics {
+            tracelimit::info_ratelimited!("EFI Diagnostics already processed, skipping");
+            return;
+        }
+        self.processed_diagnostics = true;
+
+        // Collect diagnostics logs
+        let mut logs = Vec::new();
+        match self
+            .service
+            .diagnostics
+            .process_diagnostics(gpa, gm, &mut logs)
+        {
+            Ok(_) => {
+                // Print the logs to the trace log
+                for log in logs.iter() {
+                    tracing::info!(
+                        "EFI Diagnostics: Debug Level: {}, Timestamp: {}, Phase: {}, Message: {}",
```

Comment:
Don't use string interpolation on `tracing` events, use associated key value pairs. It makes the data easier to post-process.